Fix massive spill files for StringView/BinaryView columns

EeshanBembi opened the pull request:
Add garbage collection for StringView and BinaryView arrays before spilling to disk. This prevents sliced arrays from carrying their entire original buffers when written to spill files.

Changes:
- Add gc_view_arrays() function to apply GC on view arrays
- Integrate GC into InProgressSpillFile::append_batch()
- Use simple threshold-based heuristic (100+ rows, 10KB+ buffer size)

Fixes #19414 where GROUP BY on StringView columns created 820MB spill files instead of 33MB due to sliced arrays maintaining references to original buffers.

Testing shows 80-98% reduction in spill file sizes for typical GROUP BY workloads.

EeshanBembi committed:
Fix massive spill files for StringView/BinaryView columns

Add garbage collection for StringView and BinaryView arrays before spilling
to disk. This prevents sliced arrays from carrying their entire original
buffers when written to spill files.

Changes:
- Add gc_view_arrays() function to apply GC on view arrays
- Integrate GC into InProgressSpillFile::append_batch()
- Use simple threshold-based heuristic (100+ rows, 10KB+ buffer size)

Fixes #19414 where GROUP BY on StringView columns created 820MB spill files
instead of 33MB due to sliced arrays maintaining references to original buffers.

Testing shows 80-98% reduction in spill file sizes for typical GROUP BY workloads.

2010YOUY01 commented on datafusion/physical-plan/src/spill/mod.rs:
```diff
@@ -339,6 +339,97 @@ fn get_max_alignment_for_schema(schema: &Schema) -> usize {
+    }
+
+    if any_gc_performed {
+        Ok(RecordBatch::try_new(batch.schema(), new_columns)?)
```
nit: I think we can get rid of the `any_gc_performed` condition, and always go this branch to make it a little bit simpler

ebembi-crdb replied:
Fixed, always returning a new batch for consistency.

2010YOUY01 commented on datafusion/physical-plan/src/spill/mod.rs:
```diff
@@ -799,4 +890,426 @@ mod tests {
+            "https://produkty%2Fpulove.ru/album/login",
+        ];
+
+        let mut urls = Vec::with_capacity(200_000);
```
Yes, it would be great to make tests faster and use less memory.

2010YOUY01 commented on datafusion/physical-plan/src/spill/in_progress_spill_file.rs:
```diff
@@ -50,6 +50,7 @@ impl InProgressSpillFile {
     }
 
     /// Appends a `RecordBatch` to the spill file, initializing the writer if necessary.
+    /// Performs garbage collection on StringView/BinaryView arrays to reduce spill file size.
```
I recommend to add more comments to explain the rationale for views gc, perhaps just copy and paste from https://github.com/apache/datafusion/blob/bb9a4a7ea19896fab7da6554cf5dd103e50d8e74/datafusion/physical-plan/src/sorts/sort.rs#L483

2010YOUY01 reviewed (commented):
Thanks, it's a good idea to include compaction inside `SpillManager`

One follow-on to do is refactoring the external sort, now it's doing some compaction already outside the `SpillManager`, so with this PR it would be doing redundant compactions, I believe we should just simply remove that
https://github.com/apache/datafusion/blob/bb9a4a7ea19896fab7da6554cf5dd103e50d8e74/datafusion/physical-plan/src/sorts/sort.rs#L483

EeshanBembi committed:
Address PR review feedback for StringView/BinaryView GC

- Replace row count heuristic with 10KB memory threshold
- Improve documentation and add inline comments
- Remove redundant test_exact_clickbench_issue_19414
- Maintains 96% reduction in spill file sizes

EeshanBembi committed:
Apply cargo fmt

EeshanBembi committed:
Remove redundant StringView GC in external sort

The SpillManager now handles GC for StringView/BinaryView arrays internally
via gc_view_arrays(), making the organize_stringview_arrays() function in
external sort redundant.

Changes:
- Remove organize_stringview_arrays() call and function from sort.rs
- Use batch.clone() for early return (cheaper than creating new batch)
- Use arrow_data::MAX_INLINE_VIEW_LEN constant instead of custom constant
- Update comment in spill_manager.rs to reference gc_view_arrays()

xudong963 reviewed (commented):
The fix might also resolve the https://github.com/apache/datafusion/issues/19846

EeshanBembi commented on datafusion/physical-plan/Cargo.toml:
```diff
@@ -73,6 +73,7 @@ pin-project-lite = "^0.2.7"
 tokio = { workspace = true }
 
 [dev-dependencies]
+arrow-data = { workspace = true }
```
This is intentional, arrow-data is added as a dev-dependency because the test helper calculate_string_view_waste_ratio uses arrow_data::MAX_INLINE_VIEW_LEN to determine whether a string value would be inlined in a StringView array.

adriangb commented:
Hi @EeshanBembi just pinging here, I'd love to see this across the line, just missing one final push!

EeshanBembi committed:
fix: remove unused import Array to fix clippy

EeshanBembi committed:
fix: Improve StringView/BinaryView GC with memory-based threshold

Address review comments from PR #19444:
- Replace row count heuristic with 10KB memory threshold
- Add comprehensive documentation explaining GC rationale and mechanism
- Use direct array parameter for better type safety
- Maintain early return optimization for non-view arrays

The GC now triggers based on actual buffer memory usage rather than
row counts, providing more accurate and efficient garbage collection
for sliced StringView/BinaryView arrays during spilling.

Tests confirm 80%+ reduction in spill file sizes for pathological cases
like ClickBench (820MB -> 33MB).

cetra3 reviewed (changes requested):
There is a bit of an issue with this and interaction with spilling I've found with the external sort. 

Currently, there is a bit of code that includes the max record batch size:

https://github.com/apache/datafusion/blob/ace9cd44b7356d60e6d69d0b98ac3f5606d55507/datafusion/physical-plan/src/sorts/sort.rs#L417-L422

This is something that records the size of the record batch *before* GC, which causes issues with memory accounting.  We need some way of knowing what this value is *after* it's been GC'd.

EeshanBembi committed:
fix: address PR review comments for StringView/BinaryView GC spill

- Return post-GC sliced size from append_batch so callers use the
  correct post-GC size for memory accounting (fixes cetra3's
  CHANGES_REQUESTED: max_record_batch_size was measured pre-GC in
  sort.rs and spill_manager.rs)
- Fix incorrect comment claiming Arrow gc() is a no-op; it always
  allocates new compact buffers
- Add comment in should_gc_view_array explaining why we sum
  data_buffers directly instead of using get_buffer_memory_size()
- Enhance append_batch doc comment with GC rationale per reviewer request
- Reduce row counts in heavy GC tests

cetra3 reviewed (commented):
This looks good now except for the outstanding `get_buffer_memory_size` change

adriangb commented:
I opened https://github.com/apache/datafusion/pull/21633 to continue this work since I cannot push to this branch.

EeshanBembi commented:
> I opened #21633 to continue this work since I cannot push to this branch.

Hey @adriangb , i didn't realize this required more work, i mistakenly thought it's stuck in review phase. I can still work on it if needed!

adriangb commented:
okay yes happy to continue here! I apologize if I misunderstood as well, I took it you weren't able to work on this anymore. how would you like to proceed with the PRs? I'm happy to close mine but it does have the rebase + some reviewer feedback addressed that is probably worth preserving.

adriangb commented:
@EeshanBembi #21633 is now approved. It includes your commits so I think we can just merge that and you will get commit credit for your work. Does that sound okay?

EeshanBembi commented:
> @EeshanBembi #21633 is now approved. It includes your commits so I think we can just merge that and you will get commit credit for your work. Does that sound okay?

Sure that works, thanks!
